tests/test_gdbcontroller.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+"""
+Unit tests
+
+Run from top level directory: ./tests/test_app.py
+"""
+
+import os
+import random
+import subprocess
+from distutils.spawn import find_executable
+
+import pytest
+
+from pygdbmi.constants import USING_WINDOWS, GdbTimeoutError
+from pygdbmi.gdbcontroller import GdbController
+
+
+if USING_WINDOWS:
+    MAKE_CMD = "mingw32-make.exe"
+else:
+    MAKE_CMD = "make"
+
+
+def _get_c_program(makefile_target_name: str, binary_name: str) -> str:
+    """build c program and return path to binary"""
+    if not find_executable(MAKE_CMD):
+        raise AssertionError(
+            'Could not find executable "%s". Ensure it is installed and on your $PATH.'
+            % MAKE_CMD
+        )
+
+    SAMPLE_C_CODE_DIR = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "sample_c_app"
+    )
+    binary_path = os.path.join(SAMPLE_C_CODE_DIR, binary_name)
+    # Build C program
+    subprocess.check_output(
+        [MAKE_CMD, makefile_target_name, "-C", SAMPLE_C_CODE_DIR, "--quiet"]
+    )
+    return binary_path
+
+
+def test_controller() -> None:
+    """Build a simple C program, then run it with GdbController and verify the output is parsed
+    as expected"""
+
+    # Initialize object that manages gdb subprocess
+    gdbmi = GdbController()
+
+    c_hello_world_binary = _get_c_program("hello", "pygdbmiapp.a")
+
+    if USING_WINDOWS:
+        c_hello_world_binary = c_hello_world_binary.replace("\\", "/")
+    # Load the binary and its symbols in the gdb subprocess
+    responses = gdbmi.write(
+        "-file-exec-and-symbols %s" % c_hello_world_binary, timeout_sec=1
+    )
+
+    # Verify output was parsed into a list of responses
+    assert len(responses) != 0
+    response = responses[0]
+    assert set(response.keys()) == {"message", "type", "payload", "stream", "token"}
+
+    assert response["message"] == "thread-group-added"
+    assert response["type"] == "notify"
+    assert response["payload"] == {"id": "i1"}
+    assert response["stream"] == "stdout"
+    assert response["token"] is None
+
+    responses = gdbmi.write(["-file-list-exec-source-files", "-break-insert main"])
+    assert len(responses) != 0
+
+    responses = gdbmi.write(["-exec-run", "-exec-continue"], timeout_sec=3)
+
+    # Test GdbTimeoutError exception
+    got_timeout_exception = False
+    try:
+        gdbmi.get_gdb_response(timeout_sec=0)
+    except GdbTimeoutError:
+        got_timeout_exception = True
+    assert got_timeout_exception is True
```

Comment:
nit: can use pytest.raises https://docs.pytest.org/en/stable/reference/reference.html#pytest-raises

Reply:
Ops, yes, I didn't notice this.